client/components/migration/migration-runner.js

```diff
@@ -0,0 +1,251 @@
+            migration_state: migrationState
+        })
+    } );
+
+const stateErrorHandlingAPICall = (failedMigrationState) => () => {
+    // TODO: The error message doesn't show up anywhere. For now, update the flag in the option table so
+    // we know where it failed at.
+
+    return markMigrationStartedAPICall(failedMigrationState)();
+};
+
+/**
+ * A wrapper function to make API calls and check the API status response.
+ * Throw an exception up if the status > 400.
+ *
+ * @param {function} apiFn The API function that uses fetch.
+ * @returns {Object} The API JSON response from fetch.
+ */
+const fetchAPICall = (apiFn) => async () => {
+    const apiResponse = await apiFn();
+    const apiJSONResponse = await apiResponse.json();
+    if (apiResponse.status >= 400) {
+        throw new Error(apiJSONResponse.message || translate("Failed to setup WooCommerce Shipping. Please try again."));
+    }
+    return apiJSONResponse;
+};
+
+/**
+ * This is the implementation of the state transition diagram. Please check https://github.com/Automattic/woocommerce-services/pull/2757.
+ * The key of this object can be found in the MIGRATION_ENUM_TO_STATE_NAME_MAP above.
+ *
+ * The value is an {object} with these properties:
+ *   success: {string|null} This is the state to go to "next". We can only go to this state when callback is successful.
+ *   fail: {string|null} This is the state to go to when callback fails. All failed state immediately halt the machine, check diagram.
+ *   callback: {function|null} This is the function to run when the machine is on this state. It can make API call.
+ */
+const migrationStateTransitions = {
+    stateInit: {
+        success: 'stateInstalling',
+        fail: 'stateInit',
+        callback: fetchAPICall(markMigrationStartedAPICall(MIGRATION_STATE_ENUM.STARTED)),
+    },
+    stateErrorInit: {
+        success: 'stateInit',
+        fail: 'stateErrorInit',
+        callback: stateErrorHandlingAPICall(MIGRATION_STATE_ENUM.ERROR_STARTED),
+    },
+    stateInstalling: {
+        success: 'stateActivating',
+        fail: 'stateErrorInstalling',
+        callback: fetchAPICall(installPluginAPICall),
+    },
+    stateErrorInstalling: {
+        success: 'stateInstalling',
+        fail: 'stateErrorInstalling',
+        callback: stateErrorHandlingAPICall(MIGRATION_STATE_ENUM.ERROR_INSTALLING),
+    },
+    stateActivating: {
+        success: 'stateDeactivating', // TODO: This should be stateDBMigrating to migrate DB.
+        fail: 'stateErrorActivating',
+        callback: fetchAPICall(activatePluginAPICall),
+    },
+    stateErrorActivating: {
+        success: 'stateActivating',
+        fail: 'stateErrorActivating',
+        callback: stateErrorHandlingAPICall(MIGRATION_STATE_ENUM.ERROR_ACTIVATING),
+    },
+    stateDB: {
+        success: 'stateDeactivating',
+        fail: 'stateErrorDB',
+        callback: new Promise((resolve) => resolve({status: 200})), // TODO: DB migration
```

Comment:
TODO: We need an API to trigger the DB migration. Check https://github.com/woocommerce/shipping/issues/134#issuecomment-2192135701